### PR TITLE
Generic components function for averages calc.

### PR DESCRIPTION
**Pull request by Eleonore9**
- New namespace for functions to be used by all components
- Function to calculate values (or rates) to be projected
- Test (with testing data) to check the function

### CODE REVIEW

**Comment by tcoupland** [comment]
This looks like the kind of function we were thinking of adding into apis datasets namespace. Like rollup is. Perhaps when you get it working we can just copy it over.

**Comment by Eleonore9** [comment]
@mantree thanks for commenting! Let's discuss that later :smiley:

File: src/witan/models/dem/ccm/components_functions.clj [hunk 1]
```diff
@@ -0,0 +1,16 @@
+(ns witan.models.dem.ccm.components-functions
+  (:require [incanter.core :as i]
+            [clojure.core.matrix.dataset :as ds]))
+
+;; Calculate rates/values for alternative ways to project components of change:
+;; 1) Calculate averages for fixed values/rates or trends:
+(defn calculate-averages
+  "Takes in a dataset with historical data, a number of years to average on
+  and the jumpoff year. Returns a datasets with a new column containing the
+  averages for a specific component for all the years of interest."
```
Comment (sunnytownsend) [inline src/witan/models/dem/ccm/components_functions.clj]: maybe rephrase slightly 'averages over all the years of interest for a specific component'

File: src/witan/models/dem/ccm/schemas.clj [hunk 2]
```diff
@@ -62,3 +62,11 @@
 (def AtRiskPopnSchema
   (make-ordered-ds-schema [[:gss-code s/Str] [:sex s/Str] [:age s/Int] [:year s/Int]
                            [:popn s/Num] [:actualyear s/Int] [:actualage s/Int]]))
+
+;; For the migration component
+(def MigrationEstimatesSchema
```
Comment (sunnytownsend) [inline src/witan/models/dem/ccm/schemas.clj]: this schema is going to be applicable for all the components coming from the MYEB2_CoC.csv. I would suggest changing to ComponentMYESchema

File: src/witan/models/load_data.clj [hunk 3]
```diff
@@ -100,6 +100,16 @@
   {:column-names (apply-col-names-schema NetMigrationSchema csv-data)
    :columns (vec (apply-row-schema NetMigrationSchema csv-data))})
 
+(defmethod apply-record-coercion :migration-estimates
```
Comment (sunnytownsend) [inline src/witan/models/load_data.clj]: estimates is a little unclear - suggest changing to :migration-mye

File: src/witan/models/dem/ccm/components_functions.clj [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+            [clojure.core.matrix.dataset :as ds]
+            [witan.datasets :as wds]))
+
+;; Calculate rates/values for alternative ways to project components of change:
+;; 1) Calculate averages for fixed values/rates or trends:
+(defn calculate-averages
+  "Takes in a dataset with historical data, a number of years to average on
+  and the jumpoff year. Returns a datasets with a new column containing the
+  averages for the estimate column for all the years of interest."
+  [historical-data nb-of-years jumpoff-year]
```
Comment (sunnytownsend) [inline src/witan/models/dem/ccm/components_functions.clj:12]: Piers said he didn't really use 'jump-off year' but preferred 'base year' with the definition 'the latest year of estimates'. This may be true of other demographer users too, and plan to ask some. I welcome your thoughts, but my inclination is to stick with jump off year for now, defined as 'first year of projection', as it matches the way I'm splitting up functions and the figures I've made.

**Comment by sunnytownsend** [comment]
looking good :shipit:
